src/mlpack/core/optimizers/cne/cne_impl.hpp

```diff
@@ -0,0 +1,220 @@
+
+  // Store the number of elements in a cube slice or a matrix column.
+  elements = population.n_rows * population.n_cols;
+
+  // initializing helper variables.
+  fitnessValues.set_size(populationSize);
+  double fitness = 0;
+  double lastBestFitness = 0;
+
+  Log::Info << "CNE initialized successfully. Optimization started"
+      << std::endl;
+
+  // Find the fitness before optimization using given iterate parameters.
+  for (size_t j = 0; j < numFun; j++)
+    lastBestFitness += function.Evaluate(iterate, j);
+
+  // Iterate until maximum number of generations is obtained.
+  for (size_t gen = 1; gen <= maxGenerations; gen++)
+  {
+    // Calculating fitness values of all candidates.
+    for (size_t i = 0; i < populationSize; i++)
+    {
+       // Select a candidate and insert the parameters in the function.
+       iterate = population.slice(i);
+
+       // Find fitness of candidate.
+       for (size_t j = 0; j < numFun; j++)
+         fitness += function.Evaluate(iterate, j);
+
+       // Save fitness value of the evaluated candidate.
+       fitnessValues[i] = fitness;
+       fitness = 0;
+    }
+
+      Log::Info << "Generation number: " << gen << " best fitness = "
+          << fitnessValues.min() << std::endl;
+
+      // Create next generation of species.
+      Reproduce();
+
+      // Check for termination criteria.
+      if (tolerance != -1 && tolerance >= fitnessValues.min())
+      {
+          Log::Info << "Terminating. Given fitness criteria " << tolerance
+              << " > " << fitnessValues.min() << std::endl;
+          break;
+      }
+
+      // Check for termination criteria.
+      if (objectiveChange != -1 &&
+        (lastBestFitness - fitnessValues.min()) < objectiveChange)
+      {
+          Log::Info << "Terminating. Fitness History change "
+              << (lastBestFitness - fitnessValues.min())
+              << " < " << objectiveChange << std::endl;
+          break;
+      }
+
+      // Store the best fitness of present generation.
+      lastBestFitness = fitnessValues.min();
+  }
+
+  // Set the best candidate into the network parameters.
+  iterate = population.slice(index(0));
+
+  // Find the objective function value from the best candidate found.
+  for (size_t j = 0; j < numFun; j++)
+    fitness += function.Evaluate(iterate, j);
+
+  return fitness;
+}
+
+//! Reproduce candidates to create the next generation.
+void CNE::Reproduce()
+{
+  // Sort fitness values. Smaller fitness value means better performance.
+  index = arma::sort_index(fitnessValues);
+
+  // First parent.
+  size_t mom;
+
+  // Second parent.
+  size_t dad;
+
+  for (size_t i = numElite; i < populationSize - 1; i++)
+  {
+    // Select 2 different parents from elite group randomly [0, numElite).
+    mom = mlpack::math::RandInt(0, numElite);
+    dad = mlpack::math::RandInt(0, numElite);
+
+    // Making sure both parents are not the same.
+    if (mom == dad)
+    {
+      if (dad != numElite - 1) dad++;
```

Comment:
Pedantic style issue; you can write:

```c++
if (dad != numElite - 1)
{
  dad++;
}
else
{
  dad--;
}
```

you could also use a ternary operator:

```c++
dad = dad != numElite - 1 ? dad + 1 : dad - 1;
```